src/infinitempomatrix.jl

```diff
@@ -44,111 +44,342 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
+      mat[x, x + 1] = 1
+    end
+    middle = ITensor(copy(mat), new_ind, dag(is))
+  else
+    new_ind = Index(is.space[2:(end - 1)]; dir=dir(is), tags=new_tags)
+    middle = ITensors.BlockSparseTensor(
+      Float64,
+      undef,
+      Block{2}[Block(x, x + 1) for x in 1:length(new_ind.space)],
+      (new_ind, dag(is)),
+    )
+    for x in 1:length(new_ind.space)
+      dim_block = new_ind.space[x][2]
+      ITensors.blockview(middle, Block(x, x + 1)) .= diagm(0 => ones(dim_block))
+    end
+    middle = itensor(middle)
+  end
+  return top, middle, bottom
+end
+
+function convert_itensor_33matrix(tensor; leftdir=ITensors.In, kwargs...)
+  @assert order(tensor) == 4
+  left_ind = get(kwargs, :leftindex, nothing)
+  #Identify the different indices
+  sit = filterinds(inds(tensor); tags="Site")
+  local_sit = dag(only(filterinds(sit; plev=0)))
+  #A bit roundabout as filterinds does not accept dir
+  if isnothing(left_ind)
+    temp = uniqueinds(tensor, sit)
+    if dir(temp[1]) == leftdir
+      left_ind = temp[1]
+      right_ind = temp[2]
+    else
+      left_ind = temp[2]
+      right_ind = temp[1]
+    end
+  else
+    right_ind = only(uniqueinds(tensor, sit, left_ind))
+  end
+  left_dim = dim(left_ind)
+  right_dim = dim(right_ind)
+  #Build the local projectors.
+  left_tags = get(kwargs, :left_tags, tags(left_ind))
+  top_left, middle_left, bottom_left = build_three_projectors_from_index(
+    left_ind; tags=left_tags
+  )
+  right_tags = get(kwargs, :righ_tags, tags(right_ind))
+  top_right, middle_right, bottom_right = build_three_projectors_from_index(
+    right_ind; tags=right_tags
+  )
+
+  matrix = fill(op("Zero", local_sit), 3, 3)
+  for (idx_left, proj_left) in enumerate([top_left, middle_left, bottom_left])
+    for (idx_right, proj_right) in enumerate([top_right, middle_right, bottom_right])
+      matrix[idx_left, idx_right] = proj_left * tensor * proj_right
+    end
+  end
+  return matrix
+end
+
+function convert_itensor_3vector(
+  tensor; leftdir=ITensors.In, first=false, last=false, kwargs...
+)
+  @assert order(tensor) == 3
+  #Identify the different indices
+  sit = filterinds(inds(tensor); tags="Site")
+  local_sit = dag(only(filterinds(sit; plev=0)))
+  #A bit roundabout as filterinds does not accept dir
+  old_ind = only(uniqueinds(tensor, sit))
+  if dir(old_ind) == leftdir || last
+    new_tags = get(kwargs, :left_tags, tags(old_ind))
+    top, middle, bottom = build_three_projectors_from_index(old_ind; tags=new_tags)
+    vector = fill(op("Zero", local_sit), 3, 1)
+  else
+    new_tags = get(kwargs, :right_tags, tags(old_ind))
+    top, middle, bottom = build_three_projectors_from_index(old_ind; tags=new_tags)
+    vector = fill(op("Zero", local_sit), 1, 3)
+  end
+  for (idx, proj) in enumerate([top, middle, bottom])
+    vector[idx] = proj * tensor
+  end
+  return vector
+end
+
+#Matrix multiplication for Matrix of ITensors
+function Base.:*(A::Matrix{ITensor}, B::Matrix{ITensor})
```

Comment:
It looks like this "just works" without this definition if we overload `Base.zero(::ITensor)`:
```julia
julia> using ITensors

julia> Base.zero(T::ITensor) = false * T

julia> i, j, k = Index.((2, 2, 2))
((dim=2|id=221), (dim=2|id=645), (dim=2|id=568))

julia> A = randomITensor(i, j)
ITensor ord=2 (dim=2|id=221) (dim=2|id=645)
NDTensors.Dense{Float64, Vector{Float64}}

julia> B = randomITensor(j, k)
ITensor ord=2 (dim=2|id=645) (dim=2|id=568)
NDTensors.Dense{Float64, Vector{Float64}}

julia> [A A; A A] * [B; B]
2-element Vector{ITensor}:
 ITensor ord=2
Dim 1: (dim=2|id=221)
Dim 2: (dim=2|id=568)
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
 -1.0347223964342365  -10.334190795304709
  0.6816120568533722    6.292266667641426
 ITensor ord=2
Dim 1: (dim=2|id=221)
Dim 2: (dim=2|id=568)
NDTensors.Dense{Float64, Vector{Float64}}
 2×2
 -1.0347223964342365  -10.334190795304709
  0.6816120568533722    6.292266667641426
```

Reply:
Seems like a much simpler change.
Thanks for the advice!

Reply:
There are some definitions missing in NDTensors (contracting EmptyTensor with Diag tensor).
I had to include these definitions in the ITensors.jl file.

Else, everything seems to work with this change.

Reply:
Ok, I am correcting what I said. 

Once I remove the definitions of + and * using only the functions you proposed, sometimes some weird bugs happen due to missing promote rules. 
A minimal example, I believe related but not directly equivalent:

```julia
using ITensors
using ITensorInfiniteMPS

is = Index([QN("N", 0)=>1, QN("N", 1)=>1])

a = ITensor(is, prime(is))
b = ITensor(prime(is, 2))
c = ITensor(is, prime(is), prime(is, 2))
cbis = ITensor(0, is, prime(is), prime(is, 2))

#Crash by default, with a StackOverflow of promote_rule
a*b + c

#No crash
a*b + cbis

#No crash
c + cbis
``` 

One of the reasons appear to be (as far as I can say) that a*b is not a correct tensor if we start from empty tensors.
By playing with promote_rule, I can manage to prevent the first crash, but then the previously working expression start failing.

Reply:
Looks like a pretty subtle bug, your assessment is correct that the output type of `a * b` is a bit wrong there (the tensor type has the correct number of dimensions, but the storage type doesn't), and that seems to cause some ripple affect in `a * b + c` in type promotion. It seems like both `a * b` should be fixed, and also that kind of issue should be caught better in `a * b + c`.

Do you have an example of `Matrix{ITensor} * Matrix{ITensor}` failing with that kind of issue? That technically works for me but then issues arise if the result is used to add to other tensors because the output type is a bit incorrect:
```julia
julia> using ITensors
[ Info: Precompiling ITensors [9136182c-28ba-11e9-034c-db9fb085ebd5]

julia> i = Index([QN() => 2])
(dim=2|id=375) <Out>
 1: QN() => 2

julia> a = ITensor(i, i')
ITensor ord=2
(dim=2|id=375) <Out>
 1: QN() => 2
(dim=2|id=375)' <Out>
 1: QN() => 2
NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 2}}

julia> b = ITensor(i'')
ITensor ord=1
(dim=2|id=375)'' <Out>
 1: QN() => 2
NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 1}}

julia> x = [a a; a a] * [b b; b b]; # Work ok but has the wrong storage type

julia> x[1, 1] + ITensor(i, i', i'')
ERROR: StackOverflowError:
Stacktrace:
 [1] promote_result(#unused#::Type, #unused#::Type, #unused#::Type{NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 3}}}, #unused#::Type{NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 1}}})
   @ Base ./promotion.jl:321
 [2] promote_type
   @ ./promotion.jl:307 [inlined]
 [3] promote_result(#unused#::Type, #unused#::Type, #unused#::Type{NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 1}}}, #unused#::Type{NDTensors.EmptyStorage{NDTensors.EmptyNumber, NDTensors.BlockSparse{NDTensors.EmptyNumber, Vector{NDTensors.EmptyNumber}, 3}}})
   @ Base ./promotion.jl:321
```

Reply:
@LHerviou I have a simple fix for this issue which makes sure the contraction code outputs the correct types, so then there is no issue in subsequent operations like addition. I'll make a PR to NDTensors/ITensors directly.

I can also add the definition `Base.zero(T::ITensor) = false * T` so then with those two fixes `Matrix{ITensor} * Matrix{ITensor}` will "just work".